linesearch.go

```diff
@@ -76,19 +76,17 @@ func (ls *Linesearch) Iterate(loc *Location, xNext []float64) (EvaluationType, I
 		Derivative: projGrad,
 	}
 	if ls.Method.Finished(lsLoc) {
-		if ls.lastEvalType == FuncEvaluation && loc.Gradient != nil {
-			// We have the function value at the current location, but we don't
-			// have the gradient, so get it before announcing MajorIteration.
+		// Check if the last evaluation evaluated all fields of Location.
+		complEval := complementEval(loc, ls.lastEvalType)
```

Comment:
I think you can reduce the branching here by moving the copy before the complEval check and then putting the return inside. It's a simple if statement, but the consequences are different (finish an iteration vs. don't finish an iteration), so multiple return points feels clearer to me.